### PR TITLE
BG-10112 eth recovery additions

### PR DESCRIPTION
**Pull request by dannybitgo**
Part of: https://bitgoinc.atlassian.net/browse/BG-10112
This adds support for constructing an unsigned ETH recovery transaction.
Also adds the ability to sign 2nd signature on an ETH transasction

### CODE REVIEW

File: src/v2/coins/eth.js [hunk 1]
```diff
@@ -246,6 +246,54 @@ class Eth extends BaseCoin {
     }).call(this).asCallback(callback);
   }
 
+  /**
```
Comment (tylerlevine) [inline src/v2/coins/eth.js]: Please make sure your jsdocs adhere to the standard format.

File: src/v2/coins/eth.js [hunk 2]
```diff
@@ -246,6 +246,54 @@ class Eth extends BaseCoin {
...
+    const txPrebuild = params.txPrebuild;
+
+    if (!_.isNumber(params.signingKeyNonce)) {
+      throw new Error('must have signingKeyNonce as a parameter, and it must be a number');
+    }
+    if (_.isUndefined(params.walletContractAddress)) {
+      throw new Error('params must include walletContractAddress, but got undefined');
+    }
+
+    const signingNode = prova.HDNode.fromBase58(params.prv);
```
Comment (tylerlevine) [inline src/v2/coins/eth.js]: Please use bitgo-utxo-lib instead of prova for new code.

File: src/v2/coins/eth.js [hunk 3]
```diff
@@ -246,6 +246,54 @@ class Eth extends BaseCoin {
...
+      gasPrice: new ethUtil.BN(txPrebuild.gasPrice),
+      gasLimit: new ethUtil.BN(txPrebuild.gasLimit),
+      data: sendData,
+      spendAmount: params.recipients[0].amount
+    }
+
+    const ethTx = new EthTx(ethTxParams);
+    ethTx.sign(signingKey);
+    const txHex = ethTx.serialize().toString('hex');
+    return { txHex };
```
Comment (tylerlevine) [inline src/v2/coins/eth.js]: It's a little cleaner if you do `return { txHex: ethTx.serialize().toString('hex') };` instead.

File: src/v2/coins/eth.js [hunk 4]
```diff
@@ -329,7 +383,62 @@ class Eth extends BaseCoin {
     }
   }
 
+  getAddressNonce(address, callback) {
```
Comment (tylerlevine) [inline src/v2/coins/eth.js]: JSDocs please

File: src/v2/coins/eth.js [hunk 5]
```diff
@@ -329,7 +383,62 @@ class Eth extends BaseCoin {
...
+        // Calculate last nonce used
+        const outgoingTxs = backupKeyTxList.filter((tx) => tx.from === address);
+        nonce = outgoingTxs.length;
+      }
+
+      return nonce;
+    }).call(this).asCallback(callback);
+  }
+
+
```
Comment (tylerlevine) [inline src/v2/coins/eth.js]: Extra newline

File: src/v2/coins/eth.js [hunk 6]
```diff
@@ -329,7 +383,62 @@ class Eth extends BaseCoin {
...
+   * Helper function for recover()
+   * This transforms the unsigned transaction information into a format the BitGo offline vault expects
+   * @param txInfo
+   * @param ethTx
+   * @param userKey
+   * @param backupKey
+   * @returns {{tx: *, userKey: *, backupKey: *, coin: string, amount: string, gasPrice: string, gasLimit: string, recipients: ({address, amount}|{address: ({address, amount}|string), amount: string}|string)[]}}
+   */
+  formatForOfflineVault(txInfo, ethTx, userKey, backupKey, gasPrice, gasLimit, callback) {
+    const backupHDNode = prova.HDNode.fromBase58(backupKey);
```
Comment (tylerlevine) [inline src/v2/coins/eth.js]: prova again

File: src/v2/coins/eth.js [hunk 7]
```diff
@@ -329,7 +383,62 @@ class Eth extends BaseCoin {
...
+    const backupSigningKey = backupHDNode.getKey().getPublicKeyBuffer();
+    return co(function *() {
```
Comment (tylerlevine) [inline src/v2/coins/eth.js]: Is there a great reason you don't start the coroutine immediately, and instead do some assignments before? (These could be inside the coroutine)

Comment (dannybitgo) [reply src/v2/coins/eth.js]: not a good reason, will change for clarity

**Pull request merged**